### PR TITLE
Puts slot name -> ID mapping into the model

### LINKED ISSUE #2332: slot-dom-consumer.ts:253 Slot root has unexpected inner slot (0_6_0)
Repro:

1. open https://polymerlabs.github.io/arcs-live/shell_0_6_0/web-shell/
2. type 'restaurant' in search
3. choose '[demo] find rest...'
4. return to launcher
5. choose '[demo] find rest...' a second time

Should produce numerous `unexpected inner slot` warnings.

### PR DESCRIPTION
**Pull request by piotrswigon**
Fixes #2332. I.e. Allows for template caching again.

Implements Step 1 from the new proposal in:
https://docs.google.com/document/d/1rSUsQQwyQPDNuOgOIrh6vdXto1FjUkWQZl_zvmfkKR0/edit

### CODE REVIEW

File: runtime/ts/particle-execution-context.ts [hunk 1]
```diff
@@ -116,16 +116,23 @@ export class ParticleExecutionContext {
...
-              content.template = this.substituteSlotNamesForIds(content.template);
-            } else {
-              content.template = Object.entries(content.template).reduce(
-                  (templateDictionary, [templateName, templateValue]) => {
-                    templateDictionary[templateName] = this.substituteSlotNamesForIds(templateValue);
-                    return templateDictionary;
-                  }, {});
+
+            content.model = {...content.model || {}};
+            this.providedSlots.forEach((slotId, slotName) => content.model[`__provided_slot.${slotName}`] = slotId);
```
Comment (shans) [inline runtime/ts/particle-execution-context.ts]: __provided_slot is cumbersome and difficult to read. Could we go with something more succinct?

Comment (piotrswigon) [reply runtime/ts/particle-execution-context.ts]: Maybe $slot, e.g. `$slot.action`? Dollar sign signifying a special field.

Comment (piotrswigon) [reply runtime/ts/particle-execution-context.ts]: I've moved to {{$localName}} in the end.

**Comment by piotrswigon** [comment]
I had to add one more thing: Adding these slotName -> slotIds mapping in the models for sub-templates (i.e. {$template: '...', models: [...]} construct). Otherwise you cannot refer to slots inside sub-templates, which breaks number of demoes.

I believe above is temporary. Once we move away from sub-IDs to [(Entity, Slot)] constructs this will disappear. I will prepare a one-pager for where I think we're going with slots and to justify moving in this direction.

**Pull request merged**